# Assignment no 69 (Voting Classifier and Pipeline Building) (15.4.23)

#### Q1. You are working on a machine learning project where you have a dataset containing numerical and categorical features. You have identified that some of the features are highly correlated and there are missing values in some of the columns. You want to build a pipeline that automates the feature engineering process and handles the missing values.

Design a pipeline that includes the following steps:

- Use an automated feature selection method to identify the important features in the dataset. 
- Create a numerical pipeline that includes the following steps: 
    a. Impute the missing values in the numerical columns using the mean of the column values. 
    b. Scale the numerical columns using standardization. 
- Create a categorical pipeline that includes the following steps: 
    a. Impute the missing values in the categorical columns using the most frequent value of the column. 
    b. One-hot encode the categorical columns. Combine the numerical and categorical pipelines using a              ColumnTransformer. 
- Use a Random Forest Classifier to build the final model. 
- Evaluate the accuracy of the model on the test dataset. 

Note! Your solution should include code snippets for each step of the pipeline, and a brief explanation of each step. You should also provide an interpretation of the results and suggest possible improvements for the pipeline.

1. First, we can use an automated feature selection method such as Recursive Feature Elimination (RFE) with a Random Forest estimator to identify the important features in the dataset.

```python
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier

# assuming X is the feature matrix and y is the target vector
selector = RFE(estimator=RandomForestClassifier(), n_features_to_select=10)
X_new = selector.fit_transform(X, y)
```

2. Next, we can create a numerical pipeline that includes two steps: imputing the missing values in the numerical columns using the mean of the column values and scaling the numerical columns using standardization.

```python
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

numerical_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])
```

3. We can also create a categorical pipeline that includes two steps: imputing the missing values in the categorical columns using the most frequent value of the column and one-hot encoding the categorical columns.

```python
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder

categorical_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('encoder', OneHotEncoder())
])
```

4. Next, we can combine the numerical and categorical pipelines using a ColumnTransformer.

```python
from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_pipeline, numerical_columns),
        ('cat', categorical_pipeline, categorical_columns)
    ])
```

5. Finally, we can use a Random Forest Classifier to build the final model and evaluate its accuracy on the test dataset.

```python
from sklearn.ensemble import RandomForestClassifier

model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier())
])

model.fit(X_train, y_train)
accuracy = model.score(X_test, y_test)
```

This pipeline automates the feature engineering process and handles the missing values. The accuracy of the model on the test dataset provides an estimate of its performance. Possible improvements for the pipeline could include trying different imputation strategies, scaling methods, or classification algorithms. Additionally, hyperparameter tuning could be performed to further improve the performance of the model.

#### Q2. Build a pipeline that includes a random forest classifier and a logistic regression classifier, and then use a voting classifier to combine their predictions. Train the pipeline on the iris dataset and evaluate its accuracy.

In [28]:
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

import warnings
warnings.filterwarnings('ignore')

In [29]:
iris = load_iris()

In [30]:
X, y = iris.data, iris.target

In [31]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [32]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [33]:
# splitting the dataset into training and test sets.

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

In [34]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((112, 4), (38, 4), (112,), (38,))

In [35]:
# creating a pipeline that includes a random forest classifier and a logistic regression classifier. 
# using a voting classifier to combine their predictions.

clf1 = RandomForestClassifier(n_estimators=50, random_state=0)
clf1

RandomForestClassifier(n_estimators=50, random_state=0)

In [36]:
clf2 = LogisticRegression(random_state=0)
clf2

LogisticRegression(random_state=0)

In [37]:
eclf = VotingClassifier(estimators=[('rf', clf1), ('lr', clf2)], voting='soft')
eclf

VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(n_estimators=50,
                                                     random_state=0)),
                             ('lr', LogisticRegression(random_state=0))],
                 voting='soft')

In [38]:
# training the pipeline on the iris dataset.

eclf.fit(X_train, y_train)

VotingClassifier(estimators=[('rf',
                              RandomForestClassifier(n_estimators=50,
                                                     random_state=0)),
                             ('lr', LogisticRegression(random_state=0))],
                 voting='soft')

In [39]:
# evaluating accuracy of voting classifier.

eclf.score(X_test, y_test).round(2)

0.97

The accuracy of our voting classifier model is 97%.